File: Samples/AppServices/cs/RandomNumberService/RandomNumberGeneratorTask.cs
using System;
using Windows.ApplicationModel.AppService;
using Windows.ApplicationModel.Background;
using Windows.Foundation.Collections;

namespace RandomNumberService
{
    public sealed class RandomNumberGeneratorTask : IBackgroundTask
    {
        BackgroundTaskDeferral serviceDeferral;
        AppServiceConnection connection;
        Random randomNumberGenerator;

        public void Run(IBackgroundTaskInstance taskInstance)
        {
            //Take a service deferral so the service isn't terminated
            serviceDeferral = taskInstance.GetDeferral();

            taskInstance.Canceled += OnTaskCanceled;

            //Initialize the random number generator
            randomNumberGenerator = new Random((int)DateTime.Now.Ticks);

            var details = taskInstance.TriggerDetails as AppServiceTriggerDetails;
            connection = details.AppServiceConnection;

            //Listen for incoming app service requests
            connection.RequestReceived += OnRequestReceived;
        }

        private void OnTaskCanceled(IBackgroundTaskInstance sender, BackgroundTaskCancellationReason reason)
        {
            if (serviceDeferral != null)
            {
                //Complete the service deferral
                serviceDeferral.Complete();
                serviceDeferral = null;
            }
        }

        async void OnRequestReceived(AppServiceConnection sender, AppServiceRequestReceivedEventArgs args)
        {
            //Get a deferral so we can use an awaitable API to respond to the message
            var messageDeferral = args.GetDeferral();

            try
            {
                var input = args.Request.Message;
                int minValue = (int)input["minvalue"];
                int maxValue = (int)input["maxvalue"];

                //Create the response
                var result = new ValueSet();
                result.Add("result", randomNumberGenerator.Next(minValue, maxValue));

                //Send the response
                await args.Request.SendResponseAsync(result);

            }
            finally
            {
                //Complete the message deferral so the platform knows we're done responding
                messageDeferral.Complete();
            }
        }
    }
}

File: Samples/AppServices/cpp/RandomNumberService/RandomNumberGeneratorTask.cpp
#include "pch.h"
#include "RandomNumberGeneratorTask.h"
#include <random>

using namespace concurrency;
using namespace Platform;
using namespace Windows::ApplicationModel::AppService;
using namespace Windows::ApplicationModel::Background;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;

namespace RandomNumberService
{
    Platform::Agile<Windows::ApplicationModel::Background::BackgroundTaskDeferral^> _serviceDeferral;
    AppServiceConnection^ _connection;
    std::mt19937 _randomNumberGenerator;
    void OnRequestReceived(AppServiceConnection^ connection, AppServiceRequestReceivedEventArgs^ args);

    void RandomNumberGeneratorTask::Run(IBackgroundTaskInstance^ taskInstance)
    {
        _randomNumberGenerator.seed(std::random_device()());

        // Take a deferral so the service isn't terminated
        _serviceDeferral = taskInstance->GetDeferral();

        taskInstance->Canceled += ref new BackgroundTaskCanceledEventHandler(this, &RandomNumberGeneratorTask::OnTaskCanceled);

        auto details = dynamic_cast<AppServiceTriggerDetails^>(taskInstance->TriggerDetails);
        _connection = details->AppServiceConnection;

        // Listen for incoming app service requests
        _connection->RequestReceived += ref new TypedEventHandler<AppServiceConnection^, AppServiceRequestReceivedEventArgs^>(OnRequestReceived);
    }

    void RandomNumberGeneratorTask::OnTaskCanceled(IBackgroundTaskInstance^ sender, BackgroundTaskCancellationReason reason)
    {
        if (_serviceDeferral != nullptr)
        {
            // Complete the service deferral
            _serviceDeferral->Complete();
            _serviceDeferral = nullptr;
        }

        delete _connection;
        _connection = nullptr;
    }

    void OnRequestReceived(AppServiceConnection^ sender, AppServiceRequestReceivedEventArgs^ args)
    {
        // Get a deferral so we can use an awaitable API to respond to the message
        auto messageDeferral = args->GetDeferral();

        auto input = args->Request->Message;
        int32 minValue = safe_cast<int32>(input->Lookup("minvalue"));
        int32 maxValue = safe_cast<int32>(input->Lookup("maxvalue"));
        std::uniform_int_distribution<std::mt19937::result_type> distribution(minValue, maxValue); // distribution in range [minValue, maxValue]
        int randomNum = distribution(_randomNumberGenerator);

        // Create the response
        auto result = ref new ValueSet();
        result->Insert("result", randomNum);

        // Send the response asynchronously
        create_task(args->Request->SendResponseAsync(result)).then([messageDeferral](AppServiceResponseStatus status)
        {
            // Complete the message deferral so the platform knows we're done responding
            messageDeferral->Complete();
        });
    }
}